### PR TITLE
Fix bug with allUSDT swaps resulting in extreme price impact due to config and filtering issues

### PR DESCRIPTION
**Pull request by p0mvn**
Fixes this:
image

Relevant thread:
https://osmosis-network.slack.com/archives/C06RT492UA0/p1719240431473179

## Change Summary

- Primary: Before attempting to do splits, we would filter out a route if it had the same pool found over many routes. We previously had a number of deeper instabilities when not filtering the pools that have slippage. For example, a split route swap over the same pool could invalidate the quote due to its own price impact. However, this filtering does not make sense for no slippage swap pools such as alloyed and transmuter.
- Secondary: Sequence of operations and router configuration issues.
   * Consider our sequence of operations
      - Find 20 candidate routes
      - Estimate quote over each
      - Sort by amount out
      - Keep top 4
      - Filter the ones that have repeating pool IDs
   - Now, imagine 2 cases.
       - Swap alloyed USDT
       - Swap kava.USDT
   - Problem: the top 4 routes between i) and ii) would differ and for i) two would have repeating pool IDs, getting filtered out from splits.
   - Solution: filter out the repeating before deciding which top 4 to keep
- Minor: Using wrong config value
   - See the "top 4 config above. It should actually be a top 3. Wrong variable passed.

## Testing

- All integration tests passed
- Added unit tests in the affected parts of the system
- Extracted new functions and unit tested for more coverage

### CODE REVIEW

File: tokens/usecase/pricing/worker/pricing_worker_test.go [hunk 1]
```diff
@@ -127,10 +127,10 @@ func (s *PricingWorkerTestSuite) TestUpdatePricesAsync() {
...
 
 func (s *PricingWorkerTestSuite) TestGetPrices_Chain_FindUnsupportedTokens() {
-	// env := os.Getenv("CI_SQS_PRICING_WORKER_TEST")
-	// if env != "true" {
-	// 	s.T().Skip("This test exists to identify which mainnet tokens are unsupported")
-	// }
+	env := os.Getenv("CI_SQS_PRICING_WORKER_TEST")
+	if env != "true" {
+		s.T().Skip("This test exists to identify which mainnet tokens are unsupported")
+	}
```
Comment (p0mvn) [inline tokens/usecase/pricing/worker/pricing_worker_test.go:133]: Note to reviewer: drive-by. Forgot to uncomment in another PR

File: tests/test_router_quote.py [hunk 2]
```diff
@@ -253,6 +253,27 @@ def validate_quote_test(self, quote, expected_amount_in_str, expected_denom_in,
...
+            for route in quote.route:
+                for pool in route.pools:
+                    pool_id = pool.id
+                    pool_data = conftest.shared_test_state.pool_by_id_map.get(str(pool_id))
+                    swap_fee = pool_data.get("swap_fees")
+
+                    if swap_fee != 0:
+                        assert False, f"Error: swap fee {swap_fee} is not charged for pool {pool_id}"
+        else:
+            assert quote.effective_fee > 0
```
Comment (p0mvn) [inline tests/test_router_quote.py]: Note to reviewer: drive-by fix to integration test flakiness

File: tests/data_service.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import requests
 
 # Endpoint URLs
-NUMIA_API_URL = 'https://data.numia-stage.osmosis.zone'
+NUMIA_API_URL = 'https://stage-proxy-data-api.osmosis-labs.workers.dev'
```
Comment (p0mvn) [inline tests/data_service.py:4]: Note to reviewer: drive-by fix to integration test flakiness. Numia stage is for some reason broken. Updaed to prod

File: sqsdomain/routable_pool.go [hunk 4]
```diff
@@ -1,41 +0,0 @@
...
-import (
-	"context"
-
-	sdk "github.com/cosmos/cosmos-sdk/types"
-
-	"github.com/osmosis-labs/osmosis/osmomath"
-	poolmanagertypes "github.com/osmosis-labs/osmosis/v25/x/poolmanager/types"
-)
-
-type RoutablePool interface {
```
Comment (p0mvn) [inline sqsdomain/routable_pool.go:12]: Note to reviewer: this was simply moved to `domain` package due to import cycle issues.

File: router/usecase/router_usecase.go [hunk 5]
```diff
@@ -345,37 +349,6 @@ func (r *routerUseCaseImpl) computeAndRankRoutesByDirectQuote(ctx context.Contex
...
-		// Keep only top routes for splits
-		routes = routes[:maxRoutes]
-		numRoutes = maxRoutes
-	}
-
-	// Convert routes sorted by amount out to routes
-	for i := 0; i < numRoutes; i++ {
-		// Update routes with the top routes
-		routes[i] = routesSortedByAmtOut[i].RouteImpl
-	}
```
Comment (p0mvn) [inline router/usecase/router_usecase.go:374]: Note to reviewer: this logic was moved to a different level of abstraction and refactored a bit for testability and to avoid redundant iterations

File: router/usecase/router_usecase.go [hunk 6]
```diff
@@ -330,9 +337,6 @@ func (r *routerUseCaseImpl) computeAndRankRoutesByDirectQuote(ctx context.Contex
 		return nil, nil, fmt.Errorf("no ranked routes found")
 	}
 
-	// Update ranked routes with filtered ranked routes
-	rankedRoutes = filterDuplicatePoolIDRoutes(rankedRoutes)
```
Comment (p0mvn) [inline router/usecase/router_usecase.go:334]: Note: moved to be earlier in the execution flow

File: router/usecase/router_usecase.go [hunk 7]
```diff
@@ -262,30 +261,38 @@ func filterDuplicatePoolIDRoutes(rankedRoutes []route.RouteImpl) []route.RouteIm
...
+	topQuote, routesWithAmtOut, err := estimateAndRankSingleRouteQuote(ctx, routes, tokenIn, r.logger)
 	if err != nil {
 		return nil, nil, fmt.Errorf("%s, tokenOutDenom (%s)", err, tokenOutDenom)
 	}
 
+	// Update ranked routes with filtered ranked routes
+	routes = filterAndConvertDuplicatePoolIDRankedRoutes(routesWithAmtOut)
+
+	// Cut routes for splits
+	routes = cutRoutesForSplits(maxSplitRoutes, routes)
```
Comment (p0mvn) [inline router/usecase/router_usecase.go:302]: Note to reviewer: core change

1. Filter duplicate routes
2. Cut for splits

As opposed to the opposite order before causing problems

File: router/usecase/router_usecase.go [hunk 8]
```diff
@@ -238,6 +232,11 @@ func filterDuplicatePoolIDRoutes(rankedRoutes []route.RouteImpl) []route.RouteIm
 		existsPoolID := false
 
 		for _, pool := range pools {
+			// Skip transmuter pools since they offer no slippage benefits.
+			if pool.GetSQSType() == domain.AlloyedTransmuter || pool.GetSQSType() == domain.TransmuterV1 {
+				continue
+			}
```
Comment (p0mvn) [inline router/usecase/router_usecase.go:246]: Note to reviewer: transmuter and alloyed should not be filtered as duplicates as they offer no slippage swaps.

File: router/usecase/optimized_routes_test.go [hunk 9]
```diff
@@ -612,6 +623,23 @@ func (s *RouterTestSuite) TestGetOptimalQuote_Mainnet() {
...
+			expectedRoutesCount: usdtOsmoExpectedRoutesHighLiq,
+		},
+		"kava.USDT for uosmo - should have the same routes as allUSDT for uosmo": {
+			tokenInDenom:  ALLUSDT,
+			tokenOutDenom: UOSMO,
+
+			amountIn: oneHundredThousandUSDValue,
+
+			expectedRoutesCount: usdtOsmoExpectedRoutesHighLiq,
+		},
```
Comment (p0mvn) [inline router/usecase/optimized_routes_test.go:642]: Note: these unit tests validate that the bug is fixed

File: domain/routable_pool.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	// TransmuterV1 is a TransmuterV1 pool type.
+	TransmuterV1
+	// GeneralizedCosmWasm is a GeneralizedCosmWasm pool type.
+	GeneralizedCosmWasm
+	// AlloyedTransmuter is an AlloyedTransmuter pool type.
+	AlloyedTransmuter
+)
+
+// RoutablePool is an interface that represents a pool that can be routed over.
+type RoutablePool interface {
```
Comment (p0mvn) [inline domain/routable_pool.go:35]: Note to reviewer: this was simply moved from `sqsdomain` package due to import cycle issues.

File: domain/routable_pool.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	StableSwap
+	// Concentrated is a Concentrated pool type.
+	Concentrated
+	// TransmuterV1 is a TransmuterV1 pool type.
+	TransmuterV1
+	// GeneralizedCosmWasm is a GeneralizedCosmWasm pool type.
+	GeneralizedCosmWasm
+	// AlloyedTransmuter is an AlloyedTransmuter pool type.
+	AlloyedTransmuter
+)
```
Comment (p0mvn) [inline domain/routable_pool.go:32]: Note to reviewer: these had to be added to provide granularity into all of the cw pools that the routes supports. The chain pool type does not differentiate between them

**Comment by p0mvn** [comment]
Will be a retroactive review per comment on Slack

File: domain/routable_pool.go (hunk 10 above)
Comment (cryptomatictrader) [reply domain/routable_pool.go:35]: 👌

File: domain/routable_pool.go (hunk 11 above)
Comment (cryptomatictrader) [reply domain/routable_pool.go:32]: 👌

File: router/usecase/optimized_routes_test.go (hunk 9 above)
Comment (cryptomatictrader) [reply router/usecase/optimized_routes_test.go:642]: 🆗

File: router/usecase/router_usecase.go (hunk 8 above)
Comment (cryptomatictrader) [reply router/usecase/router_usecase.go:246]: 👌

File: router/usecase/router_usecase.go (hunk 7 above)
Comment (cryptomatictrader) [reply router/usecase/router_usecase.go:302]: 👍 Got it!

**Review by cryptomatictrader — COMMENTED** [review]
LGTM. Thought it is a HUGE PR but a lot of them are just renaming. Thanks!
